sshd-core/src/main/java/org/apache/sshd/common/session/helpers/AbstractSession.java

```diff
@@ -151,24 +156,34 @@ public abstract class AbstractSession extends SessionHelper {
     protected final Map<KexProposalOption, String> clientProposal = new EnumMap<>(KexProposalOption.class);
     protected final Map<KexProposalOption, String> unmodClientProposal = Collections.unmodifiableMap(clientProposal);
     protected final Map<KexProposalOption, String> negotiationResult = new EnumMap<>(KexProposalOption.class);
-    protected final Map<KexProposalOption, String> unmodNegotiationResult = Collections.unmodifiableMap(negotiationResult);
+    protected final Map<KexProposalOption, String> unmodNegotiationResult = Collections
+            .unmodifiableMap(negotiationResult);
 
     protected KeyExchange kex;
     protected Boolean firstKexPacketFollows;
     protected boolean initialKexDone;
+    protected final AtomicBoolean strictKexSignalled = new AtomicBoolean();
+    protected final AtomicInteger newKeysReceivedCount = new AtomicInteger();
+    protected final AtomicInteger newKeysSentCount = new AtomicInteger();
+
     /**
      * Holds the current key exchange state.
      */
     protected final AtomicReference<KexState> kexState = new AtomicReference<>(KexState.UNKNOWN);
     protected final AtomicReference<DefaultKeyExchangeFuture> kexFutureHolder = new AtomicReference<>(null);
 
-    // The kexInitializedFuture is fulfilled when this side (client or server) has prepared its own proposal. Access is
+    // The kexInitializedFuture is fulfilled when this side (client or server)
+    // has prepared its own proposal. Access is
     // synchronized on kexState.
     protected DefaultKeyExchangeFuture kexInitializedFuture;
 
     /*
      * SSH packets encoding / decoding support
      */
+    /** Input packet sequence number. */
+    protected long seqi;
+    /** Output packet sequence number. */
+    protected long seqo;
```

Comment:
Unfortunatelly zour code does touch unrelated things, it would be good not to have to review them. Besides that it is relevant to the patch because you introduced a synchronized and it was not clear to me if it’s the right one (since it looks like the increase of those variables do not know about it).